File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        // bellman ford
        // loop k+1 times, for each loop
        // if not inf, check if src to dst can ahieve lower cose

        vector<int> prices(n, INT_MAX);
        prices[src] = 0;

        for (int i = 0; i < k+1; i ++) {
            vector<int> temp = prices;
            for (auto& flight : flights) {
                int cur = flight[0];
                int next = flight[1];
                int cost = flight[2];

                // cout << "k " << i << endl;

                if (prices[cur] == INT_MAX) {
                    continue;
                }

                if (prices[cur] + cost < temp[next]) {
                    // cout << next << " " << prices[cur] + cost << endl;
                    temp[next] = prices[cur] + cost;
                }
            }
            swap(temp, prices);
        }

        return prices[dst] == INT_MAX ? -1 : prices[dst];
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        MAX = float('inf')
        price = [MAX for _ in range(n)]

        # loop k + 1 times
        price[src] = 0

        for _ in range(k+1):
            tmp = price.copy()
            # for all edges, if start is not inf, check if we can reach dst with smaller price
            for start, end, cost in flights:
                if price[start] == MAX:
                    continue;

                tmp[end] = min(tmp[end], price[start] + cost)

            price = tmp


        return -1 if price[dst] == MAX else price[dst]
        